registry/proxy/proxyblobstore.go

```diff
@@ -76,77 +93,115 @@ func (pbs *proxyBlobStore) serveLocal(ctx context.Context, w http.ResponseWriter
 	return true, pbs.localStore.ServeBlob(ctx, w, r, dgst)
 }
 
-func (pbs *proxyBlobStore) storeLocal(ctx context.Context, dgst digest.Digest) error {
-	defer func() {
-		mu.Lock()
-		delete(inflight, dgst)
-		mu.Unlock()
-	}()
+func (pbs *proxyBlobStore) doServeFromLocalStore(ctx context.Context, w http.ResponseWriter, r *http.Request, dgst digest.Digest) error {
+	_, err := pbs.localStore.Stat(ctx, dgst)
+	if err != nil {
+		return err
+	}
+
+	return pbs.localStore.ServeBlob(ctx, w, r, dgst)
 
-	var desc distribution.Descriptor
-	var err error
-	var bw distribution.BlobWriter
+}
+
+func (pbs *proxyBlobStore) IsPresentLocally(ctx context.Context, dgst digest.Digest) bool {
+	_, err := pbs.localStore.Stat(ctx, dgst)
+	// Stat can report a zero sized file here if it's checked between creation
+	// and population.  Return nil error, and continue
+	return err == nil
+}
 
-	bw, err = pbs.localStore.Create(ctx)
+func (pbs *proxyBlobStore) fetchFromRemote(dgst digest.Digest, ctx context.Context, fetcher *BlobFetch) {
+	err := pbs.doFetchFromRemote(dgst, ctx, fetcher.readableWriter)
 	if err != nil {
+		dcontext.GetLogger(ctx).Errorf("Failed to fetch layer %s, error: %s", dgst, err)
+	}
+}
+
+func (pbs *proxyBlobStore) doFetchFromRemote(dgst digest.Digest, ctx context.Context, bw storage.ReadableWriter) error {
+	if err := pbs.authChallenger.tryEstablishChallenges(ctx); err != nil {
+		bw.CancelWithError(ctx, err)
 		return err
 	}
 
-	desc, err = pbs.copyContent(ctx, dgst, bw)
+	desc, err := pbs.copyContent(ctx, dgst, bw)
+	mu.Lock()
+	delete(inflight, dgst)
+	defer mu.Unlock()
 	if err != nil {
+		bw.CancelWithError(ctx, err)
+		dcontext.GetLogger(ctx).Errorf("Error copying to storage: %s", err.Error())
 		return err
 	}
 
 	_, err = bw.Commit(ctx, desc)
 	if err != nil {
+		dcontext.GetLogger(ctx).Errorf("Error committing to storage: %s", err.Error())
 		return err
 	}
 
+	blobRef, err := reference.WithDigest(pbs.repositoryName, dgst)
+	if err != nil {
+		dcontext.GetLogger(ctx).Errorf("Error creating reference: %s", err)
+		return err
+	}
+
+	pbs.scheduler.AddBlob(blobRef, repositoryTTL)
 	return nil
 }
 
 func (pbs *proxyBlobStore) ServeBlob(ctx context.Context, w http.ResponseWriter, r *http.Request, dgst digest.Digest) error {
-	served, err := pbs.serveLocal(ctx, w, r, dgst)
-	if err != nil {
-		dcontext.GetLogger(ctx).Errorf("Error serving blob from local storage: %s", err.Error())
-		return err
-	}
+	mu.Lock()
+	fetch, ok := inflight[dgst]
+	isPresent := pbs.IsPresentLocally(ctx, dgst)
 
-	if served {
-		return nil
+	dcontext.GetLogger(ctx).Debugf("digest %s present: %t, in flight: %t, in flight object: %v", dgst, isPresent, ok, fetch)
+	isNew := !isPresent && (!ok || ok && !fetch.readableWriter.IsInProgress())
```

Comment:
I'm wondering - when we can have it in `inflight`, but it not in progress?

Reply:
There is a race condition between deleting the inflight entry and closing the writer at the end. There will be a small window where it has already been finished (not in progress) but has not yet been deleted from the inflight map.